### PR TITLE
[REVIEW] Fix exception when `groupby.std()` with `by` column containing large values

### LINKED ISSUE #5762: GroupBy `std` aggregation raises `ValueError`
When grouping by a key column that contains values that are above sqrt(max) for that column type, a `ValueError` can be thrown due to unique values overflowing into same-values, which invalidates the groupby results.

```
import pandas as pd
import numpy as np
import dask.dataframe as dd

max = np.iinfo(np.uint64).max
sqrt = max ** 0.5
series = pd.Series(np.concatenate([sqrt * np.arange(5), np.arange(35)])).astype(
    "int64"
)
df = pd.DataFrame({"x": series, "z": np.arange(40), "y": np.arange(40)})
ddf = dd.from_pandas(df, npartitions=1)
ddf.groupby("x").std().compute(scheduler="single-threaded")
```
raises
```
~/anaconda3/envs/cudf_dev/lib/python3.7/site-packages/pandas/core/internals/mana
gers.py in set_axis(self, axis, new_labels)
    153             raise ValueError(
    154                 'Length mismatch: Expected axis has {old} elements, new
'
--> 155                 'values have {new} elements'.format(old=old_len, new=new
_len))
    156
    157         self.axes[axis] = new_labels

ValueError: Length mismatch: Expected axis has 35 elements, new values have 39 e
lements
```

I've created #5737 to resolve this issue. My current solution is dumb - it uses `try` and `set` operations to only square the value columns, but not the `by` column/s. If that raises an exception, it reverts to the old method which assumes that a `Series` has been passed as the by key, which won't produce the problem. I'm open to suggestions for a smarter solution.

### PR DESCRIPTION
**Pull request by thomcom**
Fixes #5762 and https://github.com/rapidsai/cudf/issues/3649

The `dask` `std` code invokes `**2`, which introduces a surjection in groupby when the `by` column is an `int64` dtype and contains values larger than `b = np.iinfo(np.int64).max ** 0.5`. Float dtypes can also produce the surjection. By squaring the `by` column, values in pandas that are multiples of `b` above all map to zero - that is, `[b, b*2, b*3]**2 = [0, 0, 0]`. After the values in the `by` column are squared, `.groupby` is called which then produces a resulting `DataFrame` with a different size than the `.groupby` called on the original, unsquared `by` column. We exposed this issue working in `cudf`, which instead of wrapping squared large `int64` values, clamps them. Any value larger than `b` above will produce the same surjection.

This snippet reproduces the failure in `cudf`. A snippet in a cell below will produce the error in pandas.
```
import cudf
import numpy as np
import dask_cudf
dtype_max = int(np.iinfo(np.int64).max ** 0.5 + 1)
df = cudf.DataFrame({"x": [1, 1, dtype_max, dtype_max], "y": [1, 2, 3, 4]})
ddf = dask_cudf.from_cudf(df, npartitions=2)
ddf.groupby('x').std().compute()
```

See https://github.com/rapidsai/cudf/issues/3649.

Pandas reproduces the same issue whenever the groupby key contains `sqrt(dtype_max)` and multiples of it. Harder to reproduce because of `pandas` wrapping instead of clamping, but still the same issue.

The issue is caused by `df[cols] = df[cols] ** 2` in `_var_chunks` - by squaring large int64 values, a surjection is created and the results of `df.groupby` and `(df**2).groupby` have different sizes.

### CODE REVIEW

**Comment by thomcom** [comment]
After a chat with the `cudf` team I think I've identified the issue as being with `df[cols] = df[cols]**2` in the `_var_chunks` function. Updating this PR with that approach.

File: dask/dataframe/groupby.py [hunk 1]
```diff
@@ -305,11 +305,15 @@ def _var_chunk(df, *index):
 
     n = g[x.columns].count().rename(columns=lambda c: (c, "-count"))
 
-    df[cols] = df[cols] ** 2
+    try:
+        for col in set(cols).difference(set(index)):
+            df[col] = df[col] ** 2
```
Comment (mrocklin) [inline dask/dataframe/groupby.py]: Why this change?  What was going on?

File: dask/dataframe/groupby.py [hunk 2]
```diff
@@ -305,11 +305,15 @@ def _var_chunk(df, *index):
...
+    except TypeError:
+        df[cols] = df[cols] ** 2
+
     g2 = _groupby_raise_unaligned(df, by=index)
     x2 = g2.sum().rename(columns=lambda c: (c, "-x2"))
 
-    x2.index = x.index
```
Comment (mrocklin) [inline dask/dataframe/groupby.py]: Why this change?

**Comment by mrocklin** [comment]
Thank you for handling this @thomcom .  I'm not totally sure what is going on here though, and would like to ask for some more information and context.

> After a chat with the cudf team I think I've identified the issue as being with df[cols] = df[cols]**2 in the _var_chunks function. Updating this PR with that approach.

What was the result of this chat?  Can you publicize the conversation here please so that the motivation of this change is recorded?  In general Dask tends to operate in the open. 

> This PR adds a test that targets pandas DataFrames, but it resolves an issue that affects pandas and cudf DataFrames, albeit differently.

Is there still a Pandas test here?  Did things break with Pandas?  If so, a test with Pandas would be very much welcome, especially because that is what we can easily test with our CI.  

I would encourage you to also include the cudf test in the dask-cudf test suite, just to make sure that some CI system regularly checks it.

File: dask/dataframe/groupby.py (hunk 1 above)
Comment (thomcom) [reply dask/dataframe/groupby.py]: The issue is that the `by` key is being squared. For large `int64` values in particular this will produce either clamping from `cudf` or overflow with `pandas`. The failure is more obvious with `cudf` because all values greater than `np.iinfo(np.int64).max**0.5` clamp to max. With pandas, all multiples of `np.iinfo(np.in64).max**0.5` wrap back around to `0`. There are probably other surjections in pandas but I can't map them all.

File: dask/dataframe/groupby.py (hunk 2 above)
Comment (thomcom) [reply dask/dataframe/groupby.py]: This is the specific failure mode: `df.groupby(by=index)` will produce `n` groupings on `index`. `(df**2).groupby(by=index)` will produce `m<n` groupings, in pandas whenever values in `index` are multiples of `np.iinfo(np.int64).max**0.5`, and in `cudf` whenever the values in `index` are simply larger than it.

Comment (thomcom) [reply dask/dataframe/groupby.py]: `len(x2.index) != len(x.index)`, and an exception is raised. Will post a reproducer in pandas and cudf soon.

**Comment by thomcom** [comment]
```
import pandas as pd
import numpy as np
import dask_cudf
import dask.dataframe as dd
max = np.iinfo(np.uint64).max
sqrt = max**0.5
series = pd.Series(np.concatenate([sqrt*np.arange(5), np.ar
df = pd.DataFrame({"x": series, "z": np.arange(40), "y": np
ddf = dd.from_pandas(df, npartitions=1)
ddf.groupby('x').std().compute(scheduler="single-threaded")
```

reproduces the failure in `pandas`.

File: dask/dataframe/groupby.py (hunk 1 above)
Comment (thomcom) [reply dask/dataframe/groupby.py]: I did it in a `try:` because `index` could be a `cudf` or `pandas` `Series` (and maybe something else? `level`?). I'm not convinced that's the right option. Now that I figured out the Pandas reproducer I can get back to writing a better fix.

**Comment by mrocklin** [comment]
> reproduces the failure in pandas.

Great.  That's probably something that we should add as a test.

**Comment by thomcom** [comment]
I added a test specifically with that code.

File: dask/dataframe/groupby.py (hunk 1 above)
Comment (beckernick) [reply dask/dataframe/groupby.py]: @thomcom do you have a better fix in mind currently, or is that more of a TODO in your mind?

**Comment by beckernick** [comment]
@thomcom , given the added test, do you think this ready for additional review?

**Comment by thomcom** [comment]
Yup, I'll update the title!

**Comment by quasiben** [comment]
@thomcom apologies fo being late to the review.  Can you rebase/merge with master again to resolve conflicts ?

File: dask/dataframe/groupby.py (hunk 2 above)
Comment (quasiben) [reply dask/dataframe/groupby.py]: It looks like this happens when you exceed max int64.  Like the following:

```python
In [55]: df.iloc[2]                                                                                                                                                                                                                                          
Out[55]: 
x    8589934592
z             2
y             2
Name: 2, dtype: int64

In [56]: df.iloc[2]**2                                                                                                                                                                                                                                       
Out[56]: 
x    0
z    4
y    4
Name: 2, dtype: int64
```

So the resulting groupyby index will be smaller due to more values becoming 0.  However, this does make the operation numerically incorrect

**Comment by quasiben** [comment]
@mrocklin are you ok merging in after fixing conflicts ?

File: dask/dataframe/groupby.py (hunk 1 above)
Comment (thomcom) [reply dask/dataframe/groupby.py]: I'm essentially happy with it as is, but I'm giving reviewers the chance to disapprove of my `try/catch` method. :) It isn't really necessary, but try/catch can always hide other errors.

File: dask/dataframe/groupby.py (hunk 2 above)
Comment (thomcom) [reply dask/dataframe/groupby.py]: That's true, but the failure in this case is when the `by` key/s square to be larger than their maximum. Those keys don't need to be squared, they're just bins for the non-by keys which are used in the calculation of standard deviation. I think. :)

**Comment by quasiben** [comment]
merging later this afternoon unless there is objection

File: dask/dataframe/tests/test_groupby.py [hunk 3]
```diff
@@ -2329,3 +2329,32 @@ def test_groupby_split_out_multiindex(split_out, column):
     )
 
     assert_eq(ddf_result, ddf_result_so1, check_index=False)
+
+
+def test_groupby_large_ints_exception_cudf():
```
Comment (mrocklin) [inline dask/dataframe/tests/test_groupby.py]: This would be the first cudf test within the Dask test suite.  Do we want to have cudf tests in Dask itself, or should these go into the dask-cudf library within the cudf repository?

File: dask/dataframe/tests/test_groupby.py [hunk 4]
```diff
@@ -2329,3 +2329,32 @@ def test_groupby_split_out_multiindex(split_out, column):
...
+
+def test_groupby_large_ints_exception_pandas():
+    max = np.iinfo(np.uint64).max
+    sqrt = max ** 0.5
+    series = pd.Series(np.concatenate([sqrt * np.arange(5), np.arange(35)])).astype(
+        "int64"
+    )
+    df = pd.DataFrame({"x": series, "z": np.arange(40), "y": np.arange(40)})
+    ddf = dd.from_pandas(df, npartitions=1)
+    ddf.groupby("x").std().compute(scheduler="single-threaded")
```
Comment (mrocklin) [inline dask/dataframe/tests/test_groupby.py]: Should we be testing this against some known result with `assert_eq`, or is computing it sufficient?

File: dask/dataframe/tests/test_groupby.py (hunk 3 above)
Comment (thomcom) [reply dask/dataframe/tests/test_groupby.py]: There's another test just above it, `test_groupby_dropna_cdf` which I copied for my test, so it isn't the first time.

File: dask/dataframe/groupby.py (hunk 1 above)
Comment (mrocklin) [reply dask/dataframe/groupby.py]: If we succeed at squaring some of the columns, but then fail on one, it looks like we end up squaring some of them twice?

Comment (mrocklin) [reply dask/dataframe/groupby.py]: Also, if we run into errors in the try block then why don't we get those same errors in the except block?  What is different there.

Comment (mrocklin) [reply dask/dataframe/groupby.py]: Do we actually get an exception with the overflow/clamping?  Or do we just get a wrong value?

File: dask/dataframe/tests/test_groupby.py (hunk 4 above)
Comment (thomcom) [reply dask/dataframe/tests/test_groupby.py]: Hmm yes i think I can compare it with `dd.groupby"x").std()`

File: dask/dataframe/groupby.py (hunk 1 above)
Comment (TomAugspurger) [reply dask/dataframe/groupby.py]: I think the bug is on L301. What if you change

```
cols = df._get_numeric_data().columns
```

to

```python
cols = x.columns
```

after L304? Then things may work out.

File: dask/dataframe/tests/test_groupby.py (hunk 3 above)
Comment (TomAugspurger) [reply dask/dataframe/tests/test_groupby.py]: Ideally we'd have a single test that's executed against different backends. Something like

```python
pytest.mark.parametrize("backend", ["pandas", "cudf"])
def test_groupby_var_large_ints(backend):
    pd = pytest.importorskip(backend)
    series = pd.Series(...)
    ...

```

Comment (thomcom) [reply dask/dataframe/tests/test_groupby.py]: That's tricky only because there are two `cudf` related imports, `cudf` and `dask_cudf`. I can write it with an `if` statement but it won't be as clean as the example you gave. What do you think? Two tests, or one test with an `if` in it?

Comment (TomAugspurger) [reply dask/dataframe/tests/test_groupby.py]: One test, with the if condition I think.

Comment (thomcom) [reply dask/dataframe/tests/test_groupby.py]: Done, look good to merge?

**Comment by thomcom** [comment]
This is ready for another review, all requests are met!

File: dask/dataframe/groupby.py (hunk 1 above)
Comment (thomcom) [reply dask/dataframe/groupby.py]: The `try` actually handles the `set(cols).difference(set(index)):` call, as `index` could be a column or a list of columns, or it could be a `Series` (or possibly something else I am not anticipating, hence the `try`). The objective here is just to square the columns that aren't listed in `by`. If that fails, then assume that `index` is not a list of `by` columns, and square everything on L312.

The exception does not occur during the overflow/clamping, but due to the surjection causes L315 to produce a difference set of key rows for `x2` than it did for `x`. The exception is then thrown on L312 because `x.index` and `x2.index` have a different number of rows.

**Comment by thomcom** [comment]
Great suggestion @TomAugspurger, thanks! I think we are golden with that change.

**Comment by TomAugspurger** [comment]
One last nitpick: I think L301 can be deleted now: https://github.com/dask/dask/pull/5737/files#diff-6e8811cb7ed0ab69a7062d0995e1d540R301

**Comment by thomcom** [comment]
done, thanks

**Comment by TomAugspurger** [comment]
Thanks!
